### PR TITLE
Support public database

### PR DESCRIPTION
**Pull request by caiyue1993**
Re #58

### CODE REVIEW

**Comment by ianbradbury-bizzibody** [comment]
I'm getting an error when building the pull request example project.

`Ambiguous use of 'firesOnRecordCreation'` on the following line of code.

`let subscription = CKQuerySubscription(recordType: syncObject.recordType, predicate: predict, subscriptionID: IceCreamSubscription.cloudKitPublicDatabaseSubscriptionID.id, options: [.firesOnRecordCreation, .firesOnRecordUpdate, .firesOnRecordDeletion])`

From this function.

    private func createSubscriptionInPublicDatabase(on syncObject: Syncable) {
        let predict = NSPredicate(value: true)
        let subscription = CKQuerySubscription(recordType: syncObject.recordType, predicate: predict, subscriptionID: IceCreamSubscription.cloudKitPublicDatabaseSubscriptionID.id, options: [.firesOnRecordCreation, .firesOnRecordUpdate, .firesOnRecordDeletion])
        
        let notificationInfo = CKSubscription.NotificationInfo()
        notificationInfo.shouldSendContentAvailable = true // Silent Push
        
        subscription.notificationInfo = notificationInfo
        
        let createOp = CKModifySubscriptionsOperation(subscriptionsToSave: [subscription], subscriptionIDsToDelete: [])
        createOp.modifySubscriptionsCompletionBlock = { _, _, _ in
            
        }
        createOp.qualityOfService = .utility
        database.add(createOp)
    }

**Comment by rodericj** [comment]
If I'm not mistaken this gets us like 40% of the way to a viable solution for #4. Great stuff. I'll start from the latest then.

**Comment by caiyue1993** [comment]
@rodericj Glad to hear that. Indeed I've refactored the code structure and make it much more reasonable I think.

File: Example/IceCream_Example/DevelopersViewController.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            /// When developers data changes in Realm, the following code will be executed
+            /// It works like magic.
+            self.developers = developers.filter { !$0.isDeleted }
+            self.tableView.reloadData()
+        }).disposed(by: bag)
+    }
+    
+    @objc func add() {
+        let user = Person()
+        user.name = "Yue Cai"
```
Comment (rodericj) [inline Example/IceCream_Example/DevelopersViewController.swift]: What's interesting is that the first time I ran this I got "Jim" added to my Realm. I see that it's defaulted to Jim, but I don't really see how that wasn't overridden given that this is how we add the new users.

**Comment by rodericj** [comment]
I'm getting pretty deep into the implementation here but I'm wondering what the expected use is for this conditional:

```
if let token = self.notificationToken {
    try! self.realm.commitWrite(withoutNotifying: [token])
} else {
    try! self.realm.commitWrite()
}
```
I'm looking to use something similar, or repurpose this to feed in other notification tokens in the shared database implementation. 

What I've done so far is to create 3 separate database managers:
1. private (this is the standard case)
2. public (this is the case introduced here)
3. shared (this is my new case). 

It would seem as though this PR assumes you'll either use public OR private as it forces you to only use one database. What I've done is created 3 different databaseManagers that all live in parallel. The issue becomes when the realm receives a new SyncObject via user action (adding a dog or cat for instance), the notification token stored in each database manager would fire and the new object would be added to all 3 remote databases which is undesirable. 

So the question is, what is the expected logic where the notificationToken would not be set? How do we get to that point? AFAIK this is always set when we `registerLocalDatabase()`.

**Comment by caiyue1993** [comment]
Hi @rodericj , actually the situation where notificationToken would not be set will not be happen. So the code could be optimized to the following:
```
self.notificationToken.flatMap { try! self.realm.commitWrite(withoutNotifying: [$0]) }
```

Though I haven't got deep into shared database, I thought shared database is a special private database. So 

> The issue becomes when the realm receives a new SyncObject via user action (adding a dog or cat for instance), the notification token stored in each database manager would fire and the new object would be added to all 3 remote databases

this issue could be a little tough to handle. As in the current version of IceCream, we suggest using syncing data in the same scope of database per `SyncEngine`. So we may need to tweak the existing code structure to make it happen.

**Pull request merged**